src/relay/pass/to_anf.cc

```diff
@@ -0,0 +1,474 @@
+
+ private:
+  class Creator;
+};
+
+// Creator of DependencyGraph
+class DependencyGraph::Creator : private ExprVisitor {
+ public:
+  explicit Creator(common::Arena* arena)
+    : arena_(arena) {}
+
+  DependencyGraph Create(const Expr& body) {
+    this->VisitExpr(body);
+    return std::move(graph_);
+  }
+
+ private:
+  /*! \brief allocator of all the internal node object */
+  common::Arena* arena_;
+  // The output.
+  DependencyGraph graph_;
+  // Update the message stored at the node.
+  void Depend(const Expr& parent, const Expr& child) {
+    if (graph_.node_map.count(parent) == 0) {
+      graph_.node_map[parent] = arena_->make<DependencyGraph::Node>();
+      graph_.node_unmap[graph_.node_map[parent]] = parent;
+    }
+
+    if (graph_.node_map.count(child) == 0) {
+      graph_.node_map[child] = arena_->make<DependencyGraph::Node>();
+      graph_.node_unmap[graph_.node_map[child]] = child;
+    }
+
+    auto* parent_link = arena_->make<LinkNode<DependencyGraph::Node*> >();
+    parent_link->value = graph_.node_map[parent];
+    graph_.node_map[child]->output.Push(parent_link);
+
+    VisitExpr(child);
+  }
+
+  std::unordered_set<Expr, NodeHash, NodeEqual> visited_;
+
+  void VisitExpr(const Expr& e) final {
+    if (visited_.count(e) == 0) {
+      if (graph_.node_map.count(e) == 0) {
+        graph_.node_map[e] = arena_->make<DependencyGraph::Node>();
+        graph_.node_unmap[graph_.node_map[e]] = e;
+      }
+      visited_.insert(e);
+      ExprFunctor<void(const Expr&)>::VisitExpr(e);
+    }
+  }
+
+  void VisitExpr_(const CallNode* c) final {
+    Expr e = GetRef<Expr>(c);
+    Depend(e, c->op);
+    for (const auto& a : c->args) {
+      Depend(e, a);
+    }
+  }
+
+  void VisitExpr_(const TupleNode* t) final {
+    Expr e = GetRef<Expr>(t);
+    for (const auto& a : t->fields) {
+      Depend(e, a);
+    }
+  }
+
+  void VisitExpr_(const TupleGetItemNode* t) final {
+    Expr e = GetRef<Expr>(t);
+    Depend(e, t->tuple);
+  }
+
+  void VisitExpr_(const IfNode* i) final {
+    Expr e = GetRef<Expr>(i);
+    Depend(e, i->cond);
+    Depend(e, i->true_branch);
+    Depend(e, i->false_branch);
+  }
+
+  void VisitExpr_(const FunctionNode* f) final {
+    Expr e = GetRef<Expr>(f);
+    Depend(e, f->body);
+    for (const auto& p : f->params) {
+      Depend(e, p);
+    }
+  }
+
+  void VisitExpr_(const LetNode* l) final {
+    Expr e = GetRef<Expr>(l);
+    Depend(e, l->var);
+    Depend(e, l->value);
+    Depend(e, l->body);
+  }
+
+  void VisitExpr_(const VarNode* v) final { }
+
+  void VisitExpr_(const GlobalVarNode* v) final { }
+
+  void VisitExpr_(const ConstantNode* c) final { }
+
+  void VisitExpr_(const OpNode* o) final { }
+};
+
+DependencyGraph DependencyGraph::Create(common::Arena* arena, const Expr& body) {
+  return Creator(arena).Create(body);
+}
+
+Expr ToANF(const Expr& e, const Module& m, std::set<GlobalVar>* gv);
+
+struct ScopeNode;
+using Scope = std::shared_ptr<ScopeNode>;
```

Comment:
An alternative to the current algorithm is to just keep a linked list to the weak ref of the Expr, and use Visit to create the Let list during fill, instead of creating it here.